### PR TITLE
Add `address` method to `ModuleInfo` trait

### PR DESCRIPTION
**Pull request by bkolad**
This PR introduces the following changes:
1. Adds `address()` method to `ModuleInfo` trait
2. Adds `module_address` to `DispatchCall` trait
3. Converts `ModuleInfo` trait to use associated types instead of generics.

### CODE REVIEW

File: sov-modules/sov-modules-macros/src/dispatch/dispatch_query.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ impl<'a> StructDef<'a> {
 
             quote::quote!(
                 #enum_ident::#name(message)=>{
-                    let #name = <#ty as sov_modules_api::ModuleInfo::#type_generics>::new(working_set.clone());
+                    let #name = <#ty as sov_modules_api::ModuleInfo>::new(working_set.clone());
```
Comment (preston-evans98) [inline sov-modules/sov-modules-macros/src/dispatch/dispatch_query.rs:31]: (Question): Why are we removing the generics here?

File: sov-modules/sov-modules-macros/src/module_info.rs [hunk 2]
```diff
@@ -199,6 +204,20 @@ fn make_prefix_func(
...
 
+fn make_fn_address(module_ident: &proc_macro2::Ident) -> proc_macro2::TokenStream {
+    quote::quote! {
+        fn address() -> sov_modules_api::Address {
+            use sov_modules_api::Hasher;
+            let module_path = module_path!();
+
+            let mut hasher = <<Self::Context as sov_modules_api::Spec>::Hasher>::new();
+            hasher.update(module_path.as_bytes());
+            hasher.update(stringify!(#module_ident).as_bytes());
```
Comment (preston-evans98) [inline sov-modules/sov-modules-macros/src/module_info.rs:215]: Is there any way that a malicious module can cause a collision, or is there  a natural domain separator at the end of module path?

Comment (bkolad) [reply sov-modules/sov-modules-macros/src/module_info.rs]: Fixed, now we calculate it as a hash of the module the prefix (which has a domain separator).

File: sov-modules/sov-modules-macros/src/dispatch/dispatch_query.rs (hunk 1 above)
Comment (bkolad) [reply sov-modules/sov-modules-macros/src/dispatch/dispatch_query.rs:31]: `ModuleInfo` used to have generic type `ModuleInfo<C: Context>` but I changed it to the associated type, and now we don't need the generic parameter anymore.

**Review by preston-evans98 — APPROVED** [review]

**Pull request merged**